openpgp/forwarding.go

```diff
@@ -11,66 +11,154 @@ import (
+type ForwardingInstance struct {
+	ForwarderKeyId uint64
+	ForwardeeKeyId uint64
+	ProxyParameter []byte
+}
 
-	if encryptionSubKey.PrivateKey.PubKeyAlgo != packet.PubKeyAlgoECDH {
-		return nil, nil, errors.InvalidArgumentError("encryption subkey is not algorithm 18 (ECDH)")
+// NewForwardingEntity generates a new forwardee key and derives the proxy parameters from the entity e.
+// If strict, it will return an error if encryption-capable non-revoked subkeys with a wrong algorithm are found,
+// instead of ignoring them
+func (e *Entity) NewForwardingEntity(
+	name, comment, email string, config *packet.Config, strict bool,
+) (
+	forwardeeKey *Entity, instances []ForwardingInstance, err error,
+) {
+	if e.PrimaryKey.Version != 4 {
+		return nil, nil, errors.InvalidArgumentError("unsupported key version")
 	}
 
-	ecdhKey, ok := encryptionSubKey.PrivateKey.PrivateKey.(*ecdh.PrivateKey)
-	if !ok {
-		return nil, nil, errors.InvalidArgumentError("encryption subkey is not type ECDH")
+	now := config.Now()
+	i := e.PrimaryIdentity()
+	if e.PrimaryKey.KeyExpired(i.SelfSignature, now) || // primary key has expired
+		i.SelfSignature == nil || // user ID has no self-signature
+		i.SelfSignature.SigExpired(now) || // user ID self-signature has expired
+		e.Revoked(now) || // primary key has been revoked
+		i.Revoked(now) { // user ID has been revoked
+		return nil, nil, errors.InvalidArgumentError("primary key is expired")
 	}
 
+	// Generate a new Primary key for the forwardee
 	config.Algorithm = packet.PubKeyAlgoEdDSA
 	config.Curve = packet.Curve25519
+	keyLifetimeSecs := config.KeyLifetime()
 
-	forwardeeKey, err = NewEntity(name, comment, email, config)
+	forwardeePrimaryPrivRaw, err := newSigner(config)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	forwardeeEcdhKey, ok := forwardeeKey.Subkeys[0].PrivateKey.PrivateKey.(*ecdh.PrivateKey)
-	if !ok {
-		return nil, nil, goerrors.New("wrong forwarding sub key generation")
+	primary := packet.NewSignerPrivateKey(now, forwardeePrimaryPrivRaw)
+
+	forwardeeKey = &Entity{
+		PrimaryKey: &primary.PublicKey,
+		PrivateKey: primary,
+		Identities: make(map[string]*Identity),
+		Subkeys:    []Subkey{},
 	}
 
-	proxyParam, err = ecdh.DeriveProxyParam(ecdhKey, forwardeeEcdhKey)
+	err = forwardeeKey.addUserId(name, comment, email, config, now, keyLifetimeSecs)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	kdf := ecdh.KDF{
-		Version: ecdh.KDFVersionForwarding,
-		Hash: ecdhKey.KDF.Hash,
-		Cipher: ecdhKey.KDF.Cipher,
-		ReplacementFingerprint: encryptionSubKey.PublicKey.Fingerprint,
+	// Init empty instances
+	instances = []ForwardingInstance{}
+
+	// Handle all forwarder subkeys
+	for _, forwarderSubKey := range e.Subkeys {
+		// Filter flags
+		if !forwarderSubKey.Sig.FlagsValid || forwarderSubKey.Sig.FlagCertify || forwarderSubKey.Sig.FlagSign ||
+			forwarderSubKey.Sig.FlagAuthenticate || forwarderSubKey.Sig.FlagGroupKey {
```

Comment:
IMO, this is in indication that the spec is a bit too specific. If you write that it's allowed, it'll raise the question why it's allowed, and if you write that it's not allowed, it'll raise the question why not. Since it's mostly orthogonal, I would drop the requirement about the flags, and just say that this transformation only works for decryption keys.

Reply:
Since this works only for the ECDH packet I don't see a signature possible, I'll work with this.